base/intfuncs.jl

```diff
@@ -755,9 +759,17 @@ function digits!(a::AbstractVector{T}, n::Integer; base::Integer = 10) where T<:
     isempty(a) && return a
 
     if base > 0
-        for i in eachindex(a)
-            n, d = divrem(n, base)
-            a[i] = d
+        if ispow2(base) && n >= 0 && n isa Base.BitInteger
+            k = trailing_zeros(base)
+            c = base - 1
```

Comment:
You make the assumption that `base` behaves like `Base.BitIntegers`, so I suggest you write in the condition 2 lines above:  `&& base <= typemax(Int)` and then `base = Int(base)`. (We can't write `base = Int(base)` unconditionally at the top of the function, because we support bases which don't fit in an `Int`.

Reply:
If your concern is that `c` would always be an `Int` or larger, even if `base` is of a smaller type, what about instead
```julia
c = base - one(base)
```
?

Reply:
My concern is that `trailing_zeros(base)` won't necessarily be implemented for all integer types, and that `base - 1` also assumes 2-complement implementation, so seems safer to just use `Int` (or `typeof(n)` which is `Base.BitIntgers`). 
Moreover, even if `base :: BigInt` which is implementing 2-complements logic, it's wasteful in the loop below to convert everything to `BigInt` with the `&` operation.

Reply:
Done!  Note, however, that while a base larger than `typemax` is possible in `digits!` digits, it doesn't currently work in `digits` because `ndigits` internally truncates the base to `Int`:
```julia
julia> digits(9223372036854775810, base = 9223372036854775808)
ERROR: InexactError: trunc(Int64, 9223372036854775808)
Stacktrace:
 [1] throw_inexacterror(::Symbol, ::Any, ::Int128) at ./boot.jl:560
 [2] checked_trunc_sint at ./boot.jl:582 [inlined]
 [3] toInt64 at ./boot.jl:631 [inlined]
 [4] Type at ./boot.jl:710 [inlined]
 [5] ndigits0zpb(::Int128, ::Int128) at ./intfuncs.jl:455
 [6] ndigits0z(::Int128, ::Int128) at ./intfuncs.jl:520
 [7] #ndigits#322 at ./intfuncs.jl:548 [inlined]
 [8] #ndigits at ./none:0 [inlined]
 [9] #digits#325 at ./intfuncs.jl:718 [inlined]
 [10] #digits at ./none:0 [inlined]
 [11] #digits#324 at ./intfuncs.jl:714 [inlined]
 [12] (::getfield(Base, Symbol("#kw##digits")))(::NamedTuple{(:base,),Tuple{Int128}}, ::typeof(digits), ::Int128) at ./none:0
 [13] top-level scope at REPL[2]:1

julia> digits!(zeros(Int, 2), 9223372036854775810, base = 9223372036854775808)
2-element Array{Int64,1}:
 2
 1
```

Reply:
Oh right, on the other hand it works with `BigInt`. We may have to refine these "casts" in a later iteration. Maybe a future improvement here would be then to convert `base` to the smallest bit integer type ">= Int" which can represent it.